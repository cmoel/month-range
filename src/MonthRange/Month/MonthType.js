import * as R from "ramda";
import daggy from "daggy";

const isNil = R.isNil;
const lt = R.flip(R.lt);
const gt = R.flip(R.gt);
const lte = R.flip(R.lte);
const gte = R.flip(R.gte);

const Month = daggy.tagged("Month", ["year", "position"]);

Month.from = R.compose(R.apply(Month), R.props(["year", "position"]));

Month.equals = R.curry(
  (m1, m2) => R.equals(m1.year, m2.year) && R.equals(m1.position, m2.position)
);

Month.lte = R.curry((m1, m2) => {
  if (lt(m1.year, m2.year)) return true;
  return lte(m1.year, m2.year) && lte(m1.position, m2.position);
});

Month.gte = R.curry((m1, m2) => {
  if (gt(m1.year, m2.year)) return true;
  return gte(m1.year, m2.year) && gte(m1.position, m2.position);
});

Month.isBetween = R.curry((first, last, month) =>
  R.cond([
    [R.always(isNil(first)), R.F],
    [R.always(isNil(last)), Month.equals(first)],
    [R.T, R.both(Month.gte(first), Month.lte(last))],
  ])(month)
);

Month.isIncludedIn = R.curry(({fst, snd}, month) =>
  R.compose(
    R.apply(R.__, [month]),
    R.partial(Month.isBetween),
    R.sortWith([R.ascend(R.prop("year")), R.ascend(R.prop("position"))])
  )([fst, snd])
);

export default Month;
