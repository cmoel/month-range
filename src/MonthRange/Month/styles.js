import * as R from "ramda";
import glamorous from "glamorous";

const {div} = glamorous;

const borderStyle = "1px solid hsl(182, 11%, 71%)";

const colorFor = R.cond([
  [R.contains("fst"), R.always("#fff")],
  [R.contains("snd"), R.always("#fff")],
  [R.T, R.always("#333")],
]);

const backgroundColorFor = R.cond([
  [R.contains("fst"), R.always("hsl(221, 97%, 54%)")],
  [R.contains("snd"), R.always("hsl(221, 97%, 54%)")],
  [R.contains("selected"), R.always("hsl(215, 50%, 94%)")],
  [R.T, R.always("#fff")],
]);

export const Div = div(
  {
    cursor: "pointer",
    fontWeight: 200,
    lineHeight: "2.5em",
    textAlign: "center",
    userSelect: "none",
    width: "52px",
  },
  ({className, position}) => ({
    borderTop: borderStyle,
    borderLeft: borderStyle,
    borderRight: position % 6 === 0 ? borderStyle : "none",
    borderBottom: position > 6 ? borderStyle : "none",
  }),
  ({className}) => ({backgroundColor: backgroundColorFor(className)}),
  ({className}) => ({color: colorFor(className)})
);
