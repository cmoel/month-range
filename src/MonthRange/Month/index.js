import React, {Component} from "react";
import * as R from "ramda";

import {Div} from "./styles";

import MonthType from "./MonthType";

const classNameFor = ({fst, snd}, month) =>
  R.cond([
    [R.equals(fst), R.always("fst")],
    [R.equals(snd), R.always("snd")],
    [MonthType.isIncludedIn({fst, snd}), R.always("selected")],
    [R.T, R.always("")],
  ])(month);

class Month extends Component {
  render() {
    const {year, position, selectMonth, selection, children} = this.props;
    const month = MonthType.from({year, position});
    return (
      <Div
        className={classNameFor(selection, month)}
        onClick={() => selectMonth(month)}
        position={position}
      >
        {children}
      </Div>
    );
  }
}

export default Month;
