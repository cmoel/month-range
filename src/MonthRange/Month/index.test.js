import React from "react";
import Month from "./index";
import renderer from "react-test-renderer";
import {shallow} from "enzyme";

import MonthType from "./MonthType";

test("renders correctly", () => {
  const fst = undefined;
  const snd = undefined;

  const component = renderer.create(
    <Month year={2018} position={1} selectMonth={m => m} selection={{fst, snd}}>
      Jan
    </Month>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

describe("when the month is the same as fst", () => {
  test("has the className 'fst'", () => {
    const fst = MonthType.from({position: 1, year: 2018});
    const snd = undefined;

    const component = shallow(
      <Month
        year={2018}
        position={1}
        selectMonth={m => m}
        selection={{fst, snd}}
      >
        Jan
      </Month>
    );

    expect(component.hasClass("fst")).toEqual(true);
    expect(component.hasClass("snd")).toEqual(false);
    expect(component.hasClass("selected")).toEqual(false);
  });
});

describe("when the month is the same as snd", () => {
  test("has the className 'snd'", () => {
    const fst = MonthType.from({position: 7, year: 2019});
    const snd = MonthType.from({position: 1, year: 2018});

    const component = shallow(
      <Month
        year={2018}
        position={1}
        selectMonth={m => m}
        selection={{fst, snd}}
      >
        Jan
      </Month>
    );

    expect(component.hasClass("fst")).toEqual(false);
    expect(component.hasClass("snd")).toEqual(true);
    expect(component.hasClass("selected")).toEqual(false);
  });
});

describe("when the month is between fst and snd", () => {
  test("has the className 'selected'", () => {
    const fst = MonthType.from({position: 7, year: 2019});
    const snd = MonthType.from({position: 1, year: 2018});

    const component = shallow(
      <Month
        year={2018}
        position={2}
        selectMonth={m => m}
        selection={{fst, snd}}
      >
        Jan
      </Month>
    );

    expect(component.hasClass("fst")).toEqual(false);
    expect(component.hasClass("snd")).toEqual(false);
    expect(component.hasClass("selected")).toEqual(true);
  });
});

describe("when the month is outside of the fst to snd range", () => {
  test("doesn't have any selected class names", () => {
    const fst = MonthType.from({position: 7, year: 2019});
    const snd = MonthType.from({position: 1, year: 2018});

    const component = shallow(
      <Month
        year={2017}
        position={2}
        selectMonth={m => m}
        selection={{fst, snd}}
      >
        Jan
      </Month>
    );

    expect(component.hasClass("fst")).toEqual(false);
    expect(component.hasClass("snd")).toEqual(false);
    expect(component.hasClass("selected")).toEqual(false);
  });
});
