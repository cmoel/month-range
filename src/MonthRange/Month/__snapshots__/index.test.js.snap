// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.glamor-0,
[data-glamor-0] {
  cursor: pointer;
  font-weight: 200;
  line-height: 2.5em;
  text-align: center;
  user-select: none;
  width: 52px;
  border-top: 1px solid hsl(182, 11%, 71%);
  border-left: 1px solid hsl(182, 11%, 71%);
  border-right: none;
  border-bottom: none;
  background-color: #fff;
  color: #333;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
}

<div
  className="glamor-0"
  onClick={[Function]}
>
  Jan
</div>
`;
